File: lib/views/trends_page.dart
import 'package:flutter/material.dart';
import 'package:syncfusion_flutter_charts/charts.dart';
import 'dashboard_page.dart';
import 'add_sensor_page.dart';

/// A page that displays trends in energy usage and suggests power-saving actions.
///
/// This page visualizes energy usage data over time and analyzes the data
/// to identify significant trends that may indicate potential optimizations.
class TrendsPage extends StatelessWidget {
  final List<Map<String, dynamic>> sensorsData; // List of sensors data

  TrendsPage({required this.sensorsData});

  @override
  Widget build(BuildContext context) {
    // Perform analysis on sensorsData to identify trends
    final trends = _analyzeTrends();

    return Scaffold(
      appBar: AppBar(
        title: Text('Trends and Power Saving Actions'), // Title of the app bar
        actions: [
          // Add navigation buttons as needed
        ],
      ),
      body: SingleChildScrollView(
        child: Column(
          children: [
            // Visualization of energy usage data
            _buildEnergyUsageChart(),
            // Display trend analysis results
            Padding(
              padding: const EdgeInsets.all(16.0),
              child: trends.isEmpty
                  ? Center(child: Text('No trends identified yet.')) // No trends found message
                  : ListView.builder(
                shrinkWrap: true,
                physics: NeverScrollableScrollPhysics(),
                itemCount: trends.length,
                itemBuilder: (context, index) {
                  final trend = trends[index];
                  return ListTile(
                    title: Text(trend['title'] ?? 'No title'), // Display trend title
                    subtitle: Text(trend['action'] ?? 'No action'), // Display recommended action
                  );
                },
              ),
            ),
          ],
        ),
      ),
    );
  }

  /// Builds a chart visualizing energy usage over time.
  Widget _buildEnergyUsageChart() {
    List<_ChartData> chartData = [];

    // Collect data points for the chart from sensorsData
    for (var sensor in sensorsData) {
      for (var dataPoint in sensor['data_points']) {
        DateTime timestamp = DateTime.fromMillisecondsSinceEpoch(
            int.parse(dataPoint['timestamp']) * 1000); // Convert timestamp
        double energyUsage =
            double.tryParse(dataPoint['energy_usage'].toString()) ?? 0; // Parse energy usage
        chartData.add(_ChartData(timestamp, energyUsage)); // Add to chart data
      }
    }

    // Sort chart data by timestamp
    chartData.sort((a, b) => a.timestamp.compareTo(b.timestamp));

    return SfCartesianChart(
      primaryXAxis: DateTimeAxis(), // X-axis for timestamps
      primaryYAxis: NumericAxis(title: AxisTitle(text: 'Energy Usage (kWh)')), // Y-axis for energy usage
      title: ChartTitle(text: 'Energy Usage Over Time'), // Chart title
      series: <ChartSeries>[
        LineSeries<_ChartData, DateTime>(
          dataSource: chartData, // Data source for the line chart
          xValueMapper: (_ChartData data, _) => data.timestamp, // X value mapping
          yValueMapper: (_ChartData data, _) => data.energyUsage, // Y value mapping
        ),
      ],
    );
  }

  /// Analyzes trends based on the sensors data and identifies significant energy usage.
  List<Map<String, String>> _analyzeTrends() {
    List<Map<String, String>> trends = []; // List to store identified trends

    // Example trend: High energy usage in HVAC
    final hvacSensors = sensorsData.where((sensor) => sensor['category'] == 'HVAC');
    double totalHvacUsage = hvacSensors.fold(0, (sum, sensor) {
      return sum + (double.tryParse(sensor['energy_usage'].toString()) ?? 0); // Aggregate usage
    });

    // Identify if total HVAC usage exceeds a threshold
    if (totalHvacUsage > 100) {
      trends.add({
        'title': 'High HVAC energy usage detected',
        'action': 'Consider optimizing HVAC settings during off-peak hours.', // Suggested action
      });
    }

    // Additional trend analysis can be added here

    return trends; // Return the list of identified trends
  }
}

/// Class to define chart data points for energy usage.
class _ChartData {
  final DateTime timestamp; // The timestamp of the data point
  final double energyUsage; // The energy usage value at that timestamp

  _ChartData(this.timestamp, this.energyUsage); // Constructor
}

File: frontend/src/components/TrendsPage.js
import React from 'react';
// Import the React wrapper components from @syncfusion/ej2-react-charts
import {
    ChartComponent,
    SeriesCollectionDirective,
    SeriesDirective,
    Inject
} from '@syncfusion/ej2-react-charts';
// Import the chart modules from the core package @syncfusion/ej2-charts
import {
    LineSeries,
    ScatterSeries,
    DateTime,
    Legend,
    Tooltip
} from '@syncfusion/ej2-charts';

function TrendsPage({ sensorsData }) {
    // Builds the chart data by iterating over sensorsData.
    const buildEnergyUsageChartData = () => {
        const chartData = [];
        sensorsData.forEach(sensor => {
            sensor.dataPoints.forEach(dataPoint => {
                // Convert Unix epoch (in seconds) to JavaScript Date object.
                const timestamp = new Date(parseInt(dataPoint.timestamp, 10) * 1000);
                const energyUsage = parseFloat(dataPoint.energy_usage) || 0;
                chartData.push({ timestamp, energyUsage });
            });
        });
        // Sort data points by timestamp.
        chartData.sort((a, b) => a.timestamp - b.timestamp);
        return chartData;
    };

    const chartData = buildEnergyUsageChartData();

    // Define peakTimes as a subset of chartData (for example, energyUsage above a threshold)
    const threshold = 100; // adjust this value as needed
    const peakTimes = chartData.filter(data => data.energyUsage > threshold);


    // Analyzes trends based on sensorsData. In this example, we aggregate energy usage for HVAC sensors.
    const analyzeTrends = () => {
        const trends = [];
        // Filter sensors for the HVAC category.
        const hvacSensors = sensorsData.filter(sensor => sensor.category === 'HVAC');
        // Aggregate energy usage from each HVAC sensor.
        const totalHvacUsage = hvacSensors.reduce((sum, sensor) => {
            // Note: Ensure sensor.energy_usage exists; otherwise, default to 0.
            return sum + (parseFloat(sensor.energy_usage ? sensor.energy_usage.toString() : '0') || 0);
        }, 0);
        // If usage exceeds a threshold, add a trend.
        if (totalHvacUsage > 100) {
            trends.push({
                title: 'High HVAC energy usage detected',
                action: 'Consider optimizing HVAC settings during off-peak hours.',
            });
        }
        // Additional trend analyses can be added here.
        return trends;
    };

    const trends = analyzeTrends();


    return (
        <div style={{ fontFamily: 'sans-serif' }}>
            {/* App Bar */}
            <header style={{ backgroundColor: '#6200EE', color: '#FFF', padding: '16px' }}>
                <h1>Trends and Power Saving Actions</h1>
            </header>
            <div style={{ padding: '16px' }}>
                {/* Energy Usage Chart */}
                <ChartComponent
                    primaryXAxis={{ valueType: 'DateTime' }}
                    primaryYAxis={{ title: 'Energy Usage (kWh)' } }
                    title="Energy Usage Over Time"
                    tooltip={{ enable: true }}
                >
                    <Inject services={[LineSeries, ScatterSeries, DateTime, Legend, Tooltip]} />
                    <SeriesCollectionDirective>
                        <SeriesDirective
                            dataSource={chartData}
                            xName="timestamp"
                            yName="energyUsage"
                            type="Line"
                            marker={{ visible: true }}
                        />
                        <SeriesDirective
                            dataSource={peakTimes}
                            xName="timestamp"
                            yName="energyUsage"
                            type="Scatter"
                            marker={{ visible: true, shape: 'Circle' }}
                            fill="red"
                        />
                    </SeriesCollectionDirective>
                </ChartComponent>
                {/* Trend Analysis */}
                <div style={{ padding: '16px' }}>
                    {trends.length === 0 ? (
                        <div style={{ textAlign: 'center' }}>No trends identified yet.</div>
                    ) : (
                        trends.map((trend, index) => (
                            <div
                                key={index}
                                style={{
                                    marginBottom: '16px',
                                    borderBottom: '1px solid #ccc',
                                    paddingBottom: '8px',
                                }}
                            >
                                <h3>{trend.title || 'No title'}</h3>
                                <p>{trend.action || 'No action'}</p>
                            </div>
                        ))
                    )}
                </div>
            </div>
        </div>
    );
}

export default TrendsPage;
